package/libpiksi/libpiksi/src/endpoint.c

```diff
@@ -118,170 +263,710 @@ pk_endpoint_t *pk_endpoint_create(const char *endpoint, pk_endpoint_type type)
+
+    if (status & LOOP_ERROR) {
+      PK_LOG_ANNO(LOG_ERR,
+                  "status: %s; error: %s",
+                  pk_loop_describe_status(status),
+                  pk_loop_last_error(loop));
     } else {
-      /* Return error */
-      piksi_log(LOG_ERR, "error in nn_send(): %s", pk_endpoint_strerror());
-      return -1;
+      PK_LOG_ANNO(LOG_ERR, "status: %s", pk_loop_describe_status(status));
     }
+
+    return;
+  }
+
+  pk_endpoint_t *ept = (pk_endpoint_t *)context;
+  ASSERT_TRACE(ept != NULL);
+
+  client_node_t *client_node = (client_node_t *)malloc(sizeof(client_node_t));
+
+  if (client_node == NULL) {
+
+    piksi_log(LOG_WARNING, "unable to add new client, closing connection");
+    int clientfd = pk_endpoint_accept(ept);
+
+    retry_on_eintr(NESTED_FN(int, (), { return shutdown(clientfd, SHUT_RDWR); }),
+                   LOG_WARNING,
+                   "Could not shutdown client socket");
+    retry_on_eintr(NESTED_FN(int, (), { return close(clientfd); }),
+                   LOG_WARNING,
+                   "Coult not close client socket");
+
+    return;
+  }
+
+  LIST_INSERT_HEAD(&ept->client_nodes_head, client_node, entries);
+
+  client_context_t *client_context = &client_node->val;
+
+  ENDPOINT_DEBUG_LOG("new client_count: %d; path: %s", ept->client_count + 1, ept->path);
+
+  int clientfd = pk_endpoint_accept(ept);
+
+  ENDPOINT_DEBUG_LOG("new client_fd: %d", clientfd);
+
+  client_context->handle = clientfd;
+  client_context->ept = ept;
+  client_context->node = client_node;
+
+  if (fcntl(clientfd, F_SETFL, fcntl(clientfd, F_GETFL, 0) | O_NONBLOCK) < 0) {
+    PK_LOG_ANNO(LOG_WARNING, "fcntl error: %s", strerror(errno));
+  }
+
+  if (++ept->client_count > MAX_CLIENTS) {
+    piksi_log(LOG_WARNING | LOG_SBP,
+              "client count exceeding expected maximum: %zu",
+              ept->client_count);
   }
+
+  client_context->poll_handle =
+    pk_loop_poll_add(loop, clientfd, handle_client_wake, client_context);
+
+  ASSERT_TRACE(client_context->poll_handle != NULL);
 }
 
-const char *pk_endpoint_strerror(void)
+static bool retry_on_eintr(eintr_fn_t the_func, int priority, const char *error_message)
+{
+  while (the_func() != 0) {
+    if (errno == EINTR) continue;
+    piksi_log(priority, "%s: %s", error_message, pk_endpoint_strerror());
+    return false;
+  }
+  return true;
+}
+
+static int read_and_receive_common(pk_endpoint_t *pk_ept, read_handler_fn_t read_handler, void *ctx)
 {
-  return nn_strerror(errno);
+  int rc = 0;
+
+  if (!valid_socket_type_for_read(pk_ept)) {
+    PK_LOG_ANNO(LOG_ERR, "invalid socket type for read");
+    return -1;
+  }
+
+  if (pk_ept->type == PK_ENDPOINT_SUB_SERVER || pk_ept->type == PK_ENDPOINT_REP) {
+
+    int64_t counter = 0;
+    ssize_t c = read(pk_ept->wakefd, &counter, sizeof(counter));
+
+    if (c != sizeof(counter)) {
+      PK_LOG_ANNO(LOG_ERR, "invalid read size from eventfd: %zd", c);
+      return -1;
+    }
+
+    ASSERT_TRACE(counter == 1);
```

Comment:
not sure what this is doing. is counter reset to `0` somewhere after read?

Reply:
Yes, this is the behavior `eventfd` kernel objects: a write increments its internal `uint64_t` value by the value specified, and a `read()` returns the current value of that `uint64_t` then resets it to zero.

Reply:
See http://man7.org/linux/man-pages/man2/eventfd.2.html for more details.